core/jobs/batch_jobs/exp_validation_jobs.py

```diff
@@ -0,0 +1,123 @@
+
+# coding: utf-8
+#
+# Copyright 2022 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Validation Jobs for tags of exploration."""
+
+from __future__ import annotations
+
+from core.domain import exp_fetchers
+from core.jobs import base_jobs
+from core.jobs.io import ndb_io
+from core.jobs.transforms import job_result_transforms
+from core.jobs.types import job_run_result
+from core.platform import models
+
+import apache_beam as beam
+import result
+from typing import List
+
+(exp_models, ) = models.Registry.import_models([models.NAMES.exploration])
+
+
+class GetNumberOfExpHavingInvalidTagsListJob(base_jobs.JobBase):
+    """Job that returns invalid exploration tags."""
+
+    def _check_tags_are_valid(
+        self, exp: exp_models.ExplorationModel
+    ) -> result.Result[
+        exp_models.ExplorationModel, str
+    ]:
+        """Returns the Result object based on the validity of tags
+        in the exploration model.
+
+        Args:
+            exp: exp_models.ExplorationModel. The model of exploration
+                from which tags are fetched.
+
+        Returns:
+            Result[exp_models.ExplorationModel, str]. Result object that
+            contains ExplorationModel when the operation is successful
+            and str (containing error message) when invalid tags occurs.
+        """
+        tags = exp.tags
+        tags_are_valid = True
+        output_string = 'The exp of id %s contains' % exp.id
+        visited = set()
+        dup: List[str] = []
+        empty_tag = max_length_exceed_tag = 0
+
+        if len(tags) > 10:
+            tags_are_valid = False
+            output_string += ' tags length more than 10,'
+        if len(set(tags)) < len(tags):
+            tags_are_valid = False
+            for ele in tags:
+                if ele.strip() != '':
+                    if ele in visited or (visited.add(ele)):
+                        dup.append(ele)
+            if len(dup) != 0:
+                output_string += f' {len(dup)} duplicate values {dup},'
```

Comment:
I don't understand the code, especially `visited.add(ele)`, cannot this be done using set operations?